plugins/workspace-backpack/src/index.js

```diff
@@ -289,19 +391,122 @@ export class Backpack {
         'translate(' + this.left_ + ',' + this.top_ + ')');
   }
 
+  /**
+   * Returns the count of items in the backpack.
+   * @return {number} The count of items.
+   */
+  getCount() {
+    return this.contents_.length;
+  }
+
+  /**
+   * Returns backpack contents XML.
+   * @return {!Array<string>} The backpack contents.
+   */
+  getContents() {
+    // Return a shallow copy of the contents array.
+    return [...this.contents_];
+  }
+
+  /**
+   * Empties the backpack's contents. If the contents-flyout is currently open
+   * it will be closed.
+   */
+  empty() {
+    if (!this.getCount()) {
+      return;
+    }
+    this.contents_.length = 0;
+    // TODO: Fire UI event for Backpack content change.
+    this.close();
+  }
+
+  /**
+   * Handles a block drop on this backpack.
+   * @param {!Blockly.BlockSvg} block The block being dropped on the backpack.
+   */
+  handleBlockDrop(block) {
+    const blockXml = Blockly.Xml.blockToDom(block);
+    this.addItem(cleanBlockXML(blockXml));
+  }
+
+  /**
+   * Adds item to backpack.
+   * @param {string} item Text representing the XML tree of a block to add,
+   *     cleaned of all unnecessary attributes.
+   */
+  addItem(item) {
+    if (this.contents_.indexOf(item) !== -1) {
+      return;
+    }
+
+    this.contents_.unshift(item);
+    while (this.contents_.length > this.maxItems_) {
+      this.contents_.pop();
+    }
+    // TODO: Fire UI event for Backpack content change.
+  }
+
+  /**
+   * Sets backpack contents XML.
+   * @param {!Array<string>} contents The new backpack contents.
+   */
+  setContents(contents) {
+    this.contents_ = [...contents];
+    while (this.contents_.length > this.maxItems_) {
+      this.contents_.pop();
+    }
+    // TODO: Fire UI event for Backpack content change.
+  }
+
+  /**
+   * Merges backpack contents XML.
+   * @param {!Array<string>} contents The backpack contents to merge.
+   */
+  mergeContents(contents) {
+    contents.forEach((item) => {
+      this.addItem(item);
+    });
+  }
+
+  /**
+   * Whether the backpack is open-able.
+   * @return {boolean} Whether the backpack is open-able.
+   * @private
+   */
+  isOpenable_() {
+    return !this.isOpen();
+  }
+
+  /**
+   * Whether the backpack is open.
+   * @return {boolean} Whether the backpack is open.
+   */
+  isOpen() {
+    return this.flyout.isVisible();
+  }
+
   /**
    * Opens the backpack flyout.
    */
   open() {
-    // TODO: Implement flyout open.
+    if (!this.isOpenable_()) {
+      return;
+    }
+    const xml = this.contents_.map((text) => Blockly.Xml.textToDom(text));
+    this.flyout.show(xml);
     // TODO: Fire UI event for Backpack open.
   }
 
   /**
    * Closes the backpack flyout.
    */
   close() {
-    // TODO: Implement flyout close.
+    if (!this.isOpen()) {
```

Comment:
Should this also use isOpenable_ for consistency?

Reply:
I think `isOpen` is the better check here. When closing the flyout, it can be closed if it was open and whether it is currently openable doesn't matter. 
Changing this to `isOpenable_` could cause problem with edge cases (depending on how `isOpenable_` is overridden) if the state changes such that is isn't openable while the flyout is open and then it can't be closed.